api/src/services/opportunities_v0_1/search_opportunities.py

```diff
@@ -51,7 +52,44 @@ def _add_query_filters(stmt: Select[tuple[Any]], query: str | None) -> Select[tu
     if query is None or len(query) == 0:
         return stmt
 
-    # TODO - will implement this in https://github.com/HHS/simpler-grants-gov/issues/1455
+    ilike_query = f"%{query}%"
+
+    # Add a left join to the assistance listing table to filter by any of its values
+    stmt = stmt.outerjoin(
+        OpportunityAssistanceListing,
+        Opportunity.opportunity_id == OpportunityAssistanceListing.opportunity_id,
+    )
+
+    """
+    This adds the following to the inner query (assuming the query value is "example")
+
+    WHERE
+        (opportunity.opportunity_title ILIKE '%example%'
+        OR opportunity.opportunity_number ILIKE '%example%'
+        OR opportunity.agency ILIKE '%example%'
+        OR opportunity_summary.summary_description ILIKE '%example%'
+        OR opportunity_assistance_listing.assistance_listing_number = 'example'
+        OR opportunity_assistance_listing.program_title ILIKE '%example%'))
+
+    Note that SQLAlchemy escapes everything and queries are actually written like:
+
+        opportunity.opportunity_number ILIKE % (opportunity_number_1)
+    """
+    stmt = stmt.where(
+        or_(
+            # Title partial match
+            Opportunity.opportunity_title.ilike(ilike_query),
+            # Number partial match
+            Opportunity.opportunity_number.ilike(ilike_query),
+            # Agency (code) partial match
+            Opportunity.agency.ilike(ilike_query),
+            # Summary description partial match
+            OpportunitySummary.summary_description.ilike(ilike_query),
+            # assistance listing number matches exactly or program title partial match
+            OpportunityAssistanceListing.assistance_listing_number == query,
+            OpportunityAssistanceListing.program_title.ilike(ilike_query),
```

Comment:
I guess this works for now - are you planning to run the query text against opensearch in the future or something?

Reply:
Overall the structure of this file looks modular and flexible with small functions like `_add_query_filters` and `_add_filters`

Reply:
Yes, the search logic is very very simplistic on purpose at the moment. I did spend some time looking into:
- Postgres text search / tsvectors: https://www.postgresql.org/docs/current/textsearch.html
- Postgres text similarity: https://www.postgresql.org/docs/current/pgtrgm.html

The textsearch in postgres seemed to not work much better to be honest, maybe I was doing something wrong but the scores it produced generally just amounted to a `0` or `~1` value indicating it was in the column. It also didn't handle the case where you search `HHS health` intending that to be the agency + a random text search term.

Text similarity did seem quite a bit better as it allows for a bit of flexibility, but would need to investigate further.

All-in-all, the more time we spend on this, the longer it takes for us to build out an actual search index.

Reply:
@rylew1, not sure what feedback you need from design here. If it's just about what data the terms search against (`opportunity_title`, `opportunity_number`, `agency`, `summary_description`, `assistance_listing_number`, `program_title` seems like more that we even anticipated 👍) I think that's fine. At least for now; we'll learn more as we test.